compiler/rustc_middle/src/ty/consts/valtree.rs

```diff
@@ -83,6 +83,21 @@ impl<'tcx> ValTree<'tcx> {
         self.try_to_scalar_int().map(|s| s.to_target_usize(tcx))
     }
 
+    /// attempts to evaluate the given valtree to bits. can fail to evaluate in the presence of
+    /// generics (or erroneous code) or if the value can't be represented as bits (e.g. because it
+    /// contains const generic parameters or pointers).
```

Comment:
This comment is horribly outdated: This function didn't evaluate the constant before (it probably did some time ago) and now it also doesn't make sense to "evaluate" a valtree. "Evaluation" is the process that converts a generic constant like `<T as Trait>::CONST` into a concrete value, but all this function does is return the value that was already computed before.

Also a valtree can't contain generic parameters, because it is already fully evaluated. And this function also doesn't check for pointers/references (it maybe should but that's for another PR).

Maybe change this to something like:

```suggestion
    /// Attempts to extract the raw bits of the valtree.
    ///
    /// Can fail if the value can't be represented as bits (e.g. because it is an aggregate).
```